cmake/OpenCVDownload.cmake

```diff
@@ -0,0 +1,124 @@
+#    UNPACK - downloaded file will be unpacked to DESTINATION_DIR
+#    RELATIVE - if set, then URL is treated as base, and FILENAME will be appended to it
+#  Note: uses OPENCV_DOWNLOAD_DIR folder as cache, default is <opencv>/.cache
+
+#
+# Main function
+#
+
+set(OPENCV_DOWNLOAD_DIR "${OpenCV_SOURCE_DIR}/.cache" CACHE PATH "Download cache dir")
+set(OPENCV_DOWNLOAD_LOG "${OpenCV_BINARY_DIR}/CMakeDownloadLog.txt")
+
+# Init downlod cache directory and log file
+if(NOT EXISTS "${OPENCV_DOWNLOAD_DIR}")
+  file(MAKE_DIRECTORY ${OPENCV_DOWNLOAD_DIR})
+endif()
+file(WRITE "${OPENCV_DOWNLOAD_LOG}" "use_cache \"${OPENCV_DOWNLOAD_DIR}\"\n")
+
+
+function(ocv_download)
+  cmake_parse_arguments(DL "UNPACK;RELATIVE_URL" "FILENAME;HASH;URL;DESTINATION_DIR;STATUS" "" ${ARGN})
+
+  ocv_assert(DEFINED DL_FILENAME)
+  ocv_assert(DEFINED DL_HASH)
+  ocv_assert(DEFINED DL_URL)
+  ocv_assert(DEFINED DL_DESTINATION_DIR)
+
+  set(${DL_STATUS} TRUE PARENT_SCOPE)
+
+  if(DL_RELATIVE_URL)
+    set(DL_URL "${DL_URL}${DL_FILENAME}")
+  endif()
+
+  set(mode "copy")
+  if(DL_UNPACK)
+    set(mode "unpack")
+  endif()
+
+  # Log all calls to file
+  file(APPEND "${OPENCV_DOWNLOAD_LOG}" "do_${mode} \"${DL_FILENAME}\" \"${DL_HASH}\" \"${DL_URL}\" \"${DL_DESTINATION_DIR}\"\n")
+
+  # ... and to console
+  message(STATUS "Download (${mode}):")
+  message(STATUS "  ${DL_FILENAME} (${DL_HASH})")
+  message(STATUS "  ${DL_URL}")
+  message(STATUS "  ${DL_DESTINATION_DIR}")
+
+  # Copy mode: check if copy destination exists and is correct
+  if(NOT DL_UNPACK)
+    set(COPY_DESTINATION "${DL_DESTINATION_DIR}/${DL_FILENAME}")
+    if(EXISTS "${COPY_DESTINATION}")
+      file(MD5 "${COPY_DESTINATION}" target_md5)
+      if(target_md5 STREQUAL DL_HASH)
+        return()
+      endif()
+    endif()
+  endif()
+
+  # Check cache first
+  set(CACHE_CANDIDATE "${OPENCV_DOWNLOAD_DIR}/${DL_HASH}-${DL_FILENAME}")
+  if(EXISTS "${CACHE_CANDIDATE}")
+    file(MD5 "${CACHE_CANDIDATE}" target_md5)
+    if(NOT target_md5 STREQUAL DL_HASH)
+      file(REMOVE ${CACHE_CANDIDATE})
+    else()
+    endif()
+  endif()
+
+  # Download
+  if(NOT EXISTS "${CACHE_CANDIDATE}")
+    file(DOWNLOAD "${DL_URL}" "${CACHE_CANDIDATE}"
+         TIMEOUT 600 STATUS status)
+    if(NOT status EQUAL 0)
+      set(msg_level FATAL_ERROR)
+      if(DEFINED DL_STATUS)
+        set(${DL_STATUS} FALSE PARENT_SCOPE)
+        set(msg_level WARNING)
+      endif()
+      message(${msg_level} "Download failed: ${status}")
+    endif()
+
+    # Don't remove this code, because EXPECTED_MD5 parameter doesn't fail "file(DOWNLOAD)" step on wrong hash
+    file(MD5 "${CACHE_CANDIDATE}" target_md5)
+    if(NOT target_md5 STREQUAL DL_HASH)
+      set(msg_level FATAL_ERROR)
+      if(DEFINED DL_STATUS)
+        set(${DL_STATUS} FALSE PARENT_SCOPE)
+        set(msg_level WARNING)
+      endif()
+      message(${msg_level} "Hash mismatch: ${target_md5}")
+    endif()
```

Comment:
Missed bailout code. We should not work with invalid files below.
File should be removed/renamed (for future diagnostics).